source/common/config/type_to_endpoint.cc

```diff
@@ -6,96 +6,296 @@
+  // envoy::config::core::v3::ApiVersion::V3.
+  envoy::config::core::v3::ApiVersion transport_api_version_;
+
+  // This hold a name for each version, for example for "envoy.api.v2.RouteDiscoveryService".
+  //
+  // {
+  //    "V2": "envoy.api.v2.RouteDiscoveryService",
+  //    "V3": "envoy.service.route.v3.RouteDiscoveryService"
+  // }
+  std::unordered_map<envoy::config::core::v3::ApiVersion, ServiceName> names_;
+};
+
+// A ServiceVersionInfoMap holds a service's version and possible names for each available version.
+// For examples:
+//
+// Given "envoy.api.v2.RouteDiscoveryService" as the service name:
+// {
+//   "envoy.api.v2.RouteDiscoveryService": {
+//     "transport_api_version_": "V2",
+//     "names_": {
+//       "V2": "envoy.api.v2.RouteDiscoveryService",
+//       "V3": "envoy.service.route.v3.RouteDiscoveryService"
+//     }
+//   }
+// }
+//
+// And for "envoy.service.route.v3.RouteDiscoveryService":
+// {
+//   "envoy.service.route.v3.RouteDiscoveryService": {
+//     "transport_api_version_": "V3",
+//     "names_": {
+//       "V2": "envoy.api.v2.RouteDiscoveryService",
+//       "V3": "envoy.service.route.v3.RouteDiscoveryService"
+//     }
+//   }
+// }
+using ServiceVersionInfoMap = std::unordered_map<ServiceName, ServiceVersionInfo>;
+
+// This creates a ServiceToServiceVersionInfoMap, with service name (e.g.
+// "envoy.api.v2.RouteDiscoveryService") as the key.
+ServiceVersionInfoMap
+createServiceVersionInfoMap(absl::string_view service_name,
+                            envoy::config::core::v3::ApiVersion transport_api_version,
+                            const std::array<std::string, 2>& versioned_service_names) {
+  const auto key = static_cast<ServiceName>(service_name);
+  return ServiceVersionInfoMap{{
+      // ServiceName as the key.
+      key,
+
+      // ServiceVersionInfo as the value.
+      ServiceVersionInfo{
+          transport_api_version,
+          {
+              {envoy::config::core::v3::ApiVersion::V2, versioned_service_names[0]},
+              {envoy::config::core::v3::ApiVersion::V3, versioned_service_names[1]},
+          },
+      },
+  }};
+}
+
+// A method name, e.g. "envoy.api.v2.RouteDiscoveryService.StreamRoutes".
+using MethodName = std::string;
+
+// A resource type URL. For example: "type.googleapis.com/envoy.api.v2.RouteConfiguration".
+using TypeUrl = std::string;
+
+using TypeUrlVersionMap = std::unordered_map<TypeUrl, envoy::config::core::v3::ApiVersion>;
+using VersionedMethodMap = std::unordered_map<envoy::config::core::v3::ApiVersion, MethodName>;
+
+struct VersionedDiscoveryType {
+  // This holds a map of type url to its corresponding version. For example:
+  // {
+  //   "type.googleapis.com/envoy.api.v2.RouteConfiguration": "V2",
+  //   "type.googleapis.com/envoy.config.route.v3.RouteConfiguration": "V3"
+  // }
+  TypeUrlVersionMap type_url_versions_;
+
+  // Versioned discovery service RPC method fully qualified names. e.g.
```

Comment:
Again, would be good to clarify resource or transport version where possible (even if somewhat obvious here). Because, later we do tend to conflate them when making a default decision.